algebird-core/src/main/scala/com/twitter/algebird/Window.scala

```diff
@@ -0,0 +1,145 @@
+ *        private val WT: Monoid[Window[T]] = WindowMonoid[T](windowSize)
+ *        def zero = W28[T](WT.zero)
+ *        def plus(a: W28[T], b: W28[T]): W28[T] =
+ *          W28[T](WT.plus(a.window, b.window))
+ *      }
+ *  }
+ *  val elements = getElements()
+ *
+ *  val trailing90Totals =
+ *    elements
+ *      .map{ W90 }
+ *      .scanLeft(W90(0)) { (a, b) => a + b }
+ *      .map{ _.total }
+ */
+
+case class Window[T](total: T, items: Queue[T]) {
+  def size = this.items.size
+}
+
+object Window {
+  def apply[T](v: T): Window[T] = Window[T](v, Queue[T](v))
+  def from[T](ts: Traversable[T])(implicit m: WindowMonoid[T]) = m.fromTraversable(ts)
+}
+
+/**
+ * Provides a natural monoid for combining windows truncated to some window size.
+ *
+ * @param windowSize Upper limit of the number of items in a window.
+ */
+
+
+case class WindowMonoid[T](
+    windowSize: Int
+)(implicit p: Priority[Group[T], Monoid[T]])
+    extends Monoid[Window[T]] {
+  val zero =
+    p match {
+      case Priority.Preferred(g) => Window(g.zero)
+      case Priority.Fallback(m)  => Window(m.zero)
+    }
+
+  def plus(a: Window[T], b: Window[T]): Window[T] =
+    p match {
+      case Priority.Preferred(g) => plusG(a, b)(g)
+      case Priority.Fallback(m)  => plusM(a, b)(m)
+    }
+
+  def plusG(a: Window[T], b: Window[T])(implicit g: Group[T]): Window[T] =
+    if (b.items.size >= windowSize) {
+      var total: T = b.total
+      var q = b.items
+      while (q.size > windowSize) {
+        total = total - q.head
+        q = q.tail
+      }
+      Window(total, q)
+    } else {
+      // we need windowSize - b.items.size from `a`
+      var truncA = a.items
+      var totalA = a.total
+      val truncTo = windowSize - b.size
+      while (truncA.size > truncTo) {
+        totalA = totalA - truncA.head
+        truncA = truncA.tail
+      }
+      val items = truncA ++ b.items
+      val total = totalA + b.total
+      Window(total, items)
+    }
+
+  def plusM(a: Window[T], b: Window[T])(implicit m: Monoid[T]): Window[T] =
+    if (b.items.size >= windowSize) {
+      val items = b.items.takeRight(windowSize)
+      val total = m.sum(items)
+      Window(total, items)
+    } else {
+      // we need windowSize - b.items.size from `a`
+      val fromA = a.items.takeRight(windowSize - b.items.size)
+      val items = fromA ++ b.items
+      val total = m.sum(fromA) + b.total
+      Window(total, items)
+    }
+
+  def fromTraversable(ts: Traversable[T]): Window[T] = {
+    val monT: Monoid[T] = p.join
+    val right = ts.toList.takeRight(windowSize)
+    val total = monT.sum(right)
+    Window(total, Queue(right: _*))
+  }
+
+  override def sumOption(ws: TraversableOnce[Window[T]]): Option[Window[T]] = {
+    if(ws.isEmpty) None
+    else {
+      val it = ws.toIterator
+      var queue = Queue[T]()
+      while (it.hasNext) {
+        queue = (queue ++ it.next.items).takeRight(windowSize)
+      }
+      Some(fromTraversable(queue))
```

Comment:
Using fromTraversable here is ignoring that you already have a queue. I think summing the queue and reusing it will be a significant performance improvement